CA-648 Abstract out Fence JWT Token Storage & Locking

wchamber opened the pull request:
Pull out the FenceTokenVendingMachine's logic for interacting with Datastore into a separate class, FenceTokenStorage.

This will allow us to inject a stub for unit testing when we cannot use real Datastore or an emulator. We can no longer rely on testbed's datastore stub as it is not available in python 3 app engine.

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [x] Get two thumbsworth of review and PO signoff if necessary. Documentation PRs only need 1 thumb.
- [x] Verify all tests go green
- [x] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [x] Test this change deployed correctly and works on dev environment after deployment

Commit:
Add a wrapper around memcache to allow that dependency to be injected.

Also add LocalCacheApi for an in-memory implementation of the add/get memcache API used by Bond. This will help us migrate to a different cache implementation.

Commit:
Add unittests for MemcacheApi by making base class CacheApiTest.

Commit:
First pass abstracting out an API for locked storage of a fence token. Needs documentation, better names, tests for DatastoreLockedStorage. Consider if tests in fence_token_vending_test are duplicate now.

Commit:
Rename to 'FenceTokenStorage' and add tests.

Commit:
Add unittests for MemcacheApi by making base class CacheApiTest.

wchamber committed:
Fix FakeFenceTokenStorage to match changes to FenceTokenStorage API.

wchamber committed:
Remove duplicate 'cache' in comment.

wchamber committed:
Merge branch 'wc-CA-648' of https://github.com/DataBiosphere/bond into wc-CA-648

wchamber committed:
Merge branch 'wc-CA-648' into wc-CA-648-datastore-fence-token

wchamber commented on fence_token_storage.py:
```diff
@@ -0,0 +1,157 @@
+            return self._lock_fence_service_account(fsa_key)
+        # We expect a transaction failure when someone else acquires the lock instead of us. That's fine, it's just
+        # a different way we could fail to acquire the lock.
+        except TransactionFailedError:
```
I scoped down this exception to a transaction failure, removing the old TODO

dvoet replied:
did you verify this is correct? I think part of the problem was that is was unclear which are the exceptions that should be handled

wchamber replied:
That was my reading of the docs here[0] but I did not try to provoke a transaction error. Looking some more, it sounds pretty ambiguous in the note[1], so I'll go back to `except:`

[0] https://cloud.google.com/appengine/docs/standard/python/datastore/transactions#what_can_be_done_in_a_transaction
[1] https://cloud.google.com/appengine/docs/standard/python/datastore/transactions#using_transactions

wchamber committed:
Change condition for ServiceAccountNotUpdatedException to raise that function if there is no expired_at message, i.e. it's still locked, instead of only if the updated account does not have a valid expires_at value.
Before, if the update lock expired during the _wait_for_update loop, we could return the FenceServiceAccount without a received key_json, as the update locked FenceServiceAccount does not have any value for expires_at. Now, the expired update locked value raises the exception.

wchamber commented on fence_token_storage.py:
```diff
@@ -0,0 +1,156 @@
+
+        else:
+            fence_service_account = self._wait_for_update(fsa_key)
+            if not fence_service_account.expires_at or fence_service_account.expires_at < datetime.datetime.now():
```
Change condition for ServiceAccountNotUpdatedException to raise that function if there is no expired_at message, i.e. it's still locked, instead of only if the updated account does not have a valid expires_at value.
Before, if the update lock expired during the _wait_for_update loop, we could return the FenceServiceAccount without a received key_json, as the update locked FenceServiceAccount does not have any value for expires_at. Now, the expired update locked value raises the exception.

wchamber committed:
Merge branch 'develop' into wc-CA-648

wchamber committed:
Merge branch 'wc-CA-648' into wc-CA-648-datastore-fence-token

wchamber committed:
Go back to catching all exceptions as failures to acquire the locks. Add some more comments on how transactions can fail even when they succeed...

gpolumbo-broad commented:
Since you're doing some rewriting here anyway, I would suggest a rename of `FenceTokenStorage` to something like, `ProviderTokenStorage`.  The reason we have anything named Fence in the code today is because when we were first writing this stuff, there was one and only one known provider, Fence.  Then we got a second one, which was just another instance of Fence, but technically, these Providers don't _need_ to be Fences.  In general in the code, any classes or variables we have named `Fence` we can probably rename to just `Provider` and it will hopefully be a little clearer and less confusing.

gpolumbo-broad commented on fence_token_storage.py:
```diff
@@ -0,0 +1,159 @@
+        return fence_service_account.key_json if fence_service_account else None
+
+    def get_or_create(self, fsa_key, prep_key_fn, fence_fetch_fn):
+        """Retrieve the stored for key, waiting as needed, or create and store the value for the key.
```
`Retrieve the stored for key`

wchamber replied:
tried to fix up the comment, PTAL

gpolumbo-broad commented on fence_token_storage.py:
```diff
@@ -0,0 +1,159 @@
+        fence_fetch_fn is not called multiple times concurrently. Arguments should work as
+        fence_fetch_fn(prep_key_fn(key)) -> returns the string fence account credentials.
+        :return returns the fence service account json key and the expiration time for that value.
+        """
```
Missing description for `:param fsa_key`

wchamber replied:
done

wchamber commented:
> Since you're doing some rewriting here anyway, I would suggest a rename of `FenceTokenStorage` to something like, `ProviderTokenStorage`...

Greg and I talked offline. s/fence/provider seems good, but we'll do it in a follow up pull so that we can do all of the moving and renaming at once. I've filed CA-657 to track this work.

gpolumbo-broad commented on fence_token_storage.py:
```diff
@@ -0,0 +1,159 @@
+            fsa_key.delete()
+        return fence_service_account.key_json if fence_service_account else None
+
+    def get_or_create(self, fsa_key, prep_key_fn, fence_fetch_fn):
```
I'm not thrilled with this name.  I am not sure I have a better suggestion though either, maybe `get_from_cache_or_provider` or something like that?   Or maybe we could just call it `get` or `retrieve` or something like that.

wchamber replied:
Yeah, I'm not wild about the name. Went with `retrieve`, which I like more than `get_or_create`, thanks! (Get sounds like a simple look up to me, and it's more than that. We're at least always retrieving it from somewhere.)

gpolumbo-broad commented on fence_token_storage.py:
```diff
@@ -0,0 +1,159 @@
+
+        return (fence_service_account.key_json, fence_service_account.expires_at)
+
+    def _fetch_service_account(self, fsa_key, prep_key_fn, fence_fetch_fn):
```
I know this is the same name as the previous version, but in the spirit of my other naming change, maybe call this one `_fetch_and_cache_service_account`

wchamber replied:
done

gpolumbo-broad commented on fence_token_storage.py:
```diff
@@ -0,0 +1,159 @@
+_FSA_KEY_LIFETIME = datetime.timedelta(days=5)
+
+
+def create_fence_service_account_key(provider_name, user_id):
```
Perhaps a nit, but I don't believe this actually persists the Key, and when I read the word `create`, I keep expecting this method to make a new object and persist it.  (Old Rails naming conventions).   I might recommend renaming this to `make_fence_service_account_key` or maybe `build_fence_service_account_key`.

wchamber replied:
Done. I'm not familiar with the Rails convention. In my java land experience, `create` is typical for a factory function, but `build` is pretty reasonable too.

wchamber committed:
Fix requested changes.
Added some comments and renamed functions s/get_or_create/retrieve and s/_fetch_service_account/_fetch_and_cache_service_account and s/create_fence_service_account_key/build_fence_service_account_key.
